extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,1263 @@
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.joda.time.DateTime;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Random;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingDeque;
+
+/**
+ * Supervisor responsible for managing the KafkaIndexTasks for a single dataSource. At a high level, the class accepts a
+ * {@link KafkaSupervisorSpec} which includes the Kafka topic and configuration as well as an ingestion spec which will
+ * be used to generate the indexing tasks. The run loop periodically refreshes its view of the Kafka topic's partitions
+ * and the list of running indexing tasks and ensures that all partitions are being read from and that there are enough
+ * tasks to satisfy the desired number of replicas. As tasks complete, new tasks are queued to process the next range of
+ * Kafka offsets.
+ */
+public class KafkaSupervisor implements Supervisor
+{
+  private static final EmittingLogger log = new EmittingLogger(KafkaSupervisor.class);
+  private static final Random RANDOM = new Random();
+  private static final long MAX_RUN_FREQUENCY_MILLIS = 1000; // prevent us from running too often in response to events
+
+  // Internal data structures
+  // --------------------------------------------------------
+
+  /**
+   * A TaskGroup is the main data structure used by KafkaSupervisor to organize and monitor Kafka partitions and
+   * indexing tasks. All the tasks in a TaskGroup should always be doing the same thing (reading the same partitions and
+   * starting from the same offset) and if [replicas] is configured to be 1, a TaskGroup will contain a single task (the
+   * exception being if the supervisor started up and discovered and adopted some already running tasks). At any given
+   * time, there should only be up to a maximum of [taskCount] actively-reading task groups (tracked in the [taskGroups]
+   * map) + zero or more pending-completion task groups (tracked in [pendingCompletionTaskGroups]).
+   */
+  private class TaskGroup
+  {
+    // This specifies the partitions and starting offsets for this task group. It is set on group creation from the data
+    // in [partitionGroups] and never changes during the lifetime of this task group, which will live until a task in
+    // this task group has completed successfully, at which point this will be destroyed and a new task group will be
+    // created with new starting offsets. This allows us to create replacement tasks for failed tasks that process the
+    // same offsets, even if the values in [partitionGroups] has been changed.
+    Map<Integer, Long> partitionOffsets = new HashMap<>();
```

Comment:
It looks like `partitionOffsets` and `tasks` aren't being reassigned, so please make them final.